### DISCUSSION TITLE
Scd40 not working

### DISCUSSION THREAD

**Discussion by bobybobybob** [arendst/Tasmota#20155] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [yes] Device used (e.g., Sonoff Basic): D1 Mini
- [yes ] Tasmota binary firmware version number used: 13.2.0
  - [ ] Pre-compiled
  - [yes ] Self-compiled
- [ ] Flashing tools used: TasmoAdmin
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
15:13:50.475 CMD: status 0
15:13:50.481 RSL: STATUS = {"Status":{"Module":18,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_D665B2","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
15:13:50.513 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://192.168.1.18:3004/data/firmwares/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:05:58","StartupUTC":"2023-12-03T14:07:52","Sleep":50,"CfgHolder":4617,"BootCount":12,"BCResetTime":"2023-12-03T14:35:06","SaveCount":23,"SaveAddress":"3FB000"}}
15:13:50.546 RSL: STATUS2 = {"StatusFWR":{"Version":"13.2.0(TasmoCompiler-esp82664M)","BuildDateTime":"2023.12.03 13:52:28","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"370/699"}}
15:13:50.567 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["caswifi",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
15:13:50.598 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":608,"Free":1436,"Heap":25,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"164020","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000407","0F80458E","04260001","00000003","00000000","D000E881","00000084","00201000","54004000","04000080"],"Drivers":"1,2,4,6,8,9,10,12,18,19,22,24,26,29,30,35,37,38,45,50,62,68","Sensors":"1,2,63,92,111","I2CDriver":"43,62,84"}}
15:13:50.631 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota-D665B2-1458","IPAddress":"192.168.1.241","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"68:C6:3A:D6:65:B2","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
15:13:50.655 RSL: STATUS6 = {"StatusMQT":{"MqttHost":"","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_D665B2","MqttUser":"DVES_USER","MqttCount":0,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
15:13:50.678 RSL: STATUS7 = {"StatusTIM":{"UTC":"2023-12-03T14:13:50","Local":"2023-12-03T15:13:50","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":"+01:00","Sunrise":"08:24","Sunset":"16:54"}}
15:13:50.695 RSL: STATUS10 = {"StatusSNS":{"Time":"2023-12-03T15:13:50"}}
15:13:50.702 RSL: STATUS11 = {"StatusSTS":{"Time":"2023-12-03T15:13:50","Uptime":"0T00:05:58","UptimeSec":358,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"Wifi":{"AP":1,"SSId":"caswifi","BSSId":"38:10:D5:43:1D:70","Channel":2,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":1,"Downtime":"0T00:00:06"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
00:00:00.673 SCD40 serial nr 0x73FE 0x4F07 0x3B6F
00:00:00.674 I2C: SCD40 found at 0x62

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Compiler 
#define USE_I2C  
#define USE_ENS16x 
#define USE_AHT2x 
#define USE_SCD40

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
To see the values of that sensor scd40

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
i see no text, no name or values of that sensor. 

### ADDITIONAL CONTEXT
_Add any other context about the problem here._
15:17:34.070 CMD: i2cscan
15:17:34.090 RSL: RESULT = {"I2CScan":"Device(s) found at 0x62"}

15:19:56.677 CMD: i2cdriver
15:19:56.681 RSL: RESULT = {"I2CDriver":"43,62,84"}

15:22:37.328 CMD: scd40cal
15:22:37.334 RSL: RESULT = {"Command":"Unknown"}

15:23:16.831 CMD: Scd40Test
15:23:16.839 RSL: RESULT = {"SCD40Test":-1}

15:24:36.459 CMD: Scd40Strt 
15:24:36.465 RSL: RESULT = {"SCD40Strt":0}

15:25:17.323 CMD: Scd40StLP 
15:25:17.329 RSL: RESULT = {"SCD40StLP":0}

I have no idea about these commands. They are mentioned on some Tasmota page where someone couldn't get the sensor to work. Otherwise I can't find any other website where these commands are listed. I just typed them in. 

I also have an ENS160+AHT2X. When I plug this in, at least these values are displayed. 

I use Raspberry Pi Zero USB Adapter 5.1V  2.5A       Current only that sensor alone.

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
I'm having SCD40 work fine without any trouble.
`"SCD40":{"CarbonDioxide":603,"eCO2":617,"Temperature":6.54,"Humidity":66.17,"DewPoint":0.67}`

WIth a D1 mini, one risk factor is that you are using a clone board with a too weak voltage regulator (that's very common), and not able to deliver the needed power for the sensors. SCD40 may need up to over 200 mA, which is more than the 150 mA D1 Mini clones have, even ignoring that the ESP8266 itself needs most of the low amount of available power. 

To be clear, this is not related to your 5V power supply, but the 3.3V power from the voltage regulator.

**Comment by bobybobybob** [comment]
What Microcontroller Board do you use? I have some other that i can test with that sensor.

**Reply by sfromis** [reply, marked as answer]
I use a board with a good 500 mA voltage regulator, which NodeMCU boards typically have. Also using it with a couple of ESP32 boards. Original branded Wemos D1 Mini should also be workable, but with clones it is a question of luck with penny-pinching vendors.

**Comment by bobybobybob** [comment]
wow, that's really crazy. External laboratory power supply 3.3V on the D1 Mini and the SCD40 now displays values. But the values on the EN160 are now all at 0 and after a few seconds everything hangs up. Only when I remove the EN160 does everything continue to work with the other sensor. If I remove the SCD40 instead, the EN160 returns to normal values.

**Reply by sfromis** [reply]
I have no experience with EN160, and with the driver being relatively new IDK how much it has been used in a mix with other sensors. Occasionally some I2C sensors or drivers have issues acting well on the I2C bus. Usually, it can be a plan to disable drivers not in use, but what you posted indicates a built witout even having "extra" drivers.

Depending on what pullup resistors are in place (typical I2C boards has some), there is a risk of multiple sensors resulting in too much pullup in total.

**Comment by bobybobybob** [comment]
Thanks again for the further advice. I soldered out the one called B22 or 822 resistor arrays  from the ENS160 and now both sensors are working fine. 

EDIT: I rejoiced too soon with the ENS160, it ran much longer but then the hang-ups came again. But no matter, the main thing is that the SCD40 works.
